reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,144 @@
+    private val scope = CoroutineScope(schedulerJob + dispatcher + CoroutineExceptionHandler { _, throwable ->
+        RxJavaPlugins.onError(throwable)
+    })
+
+    override fun scheduleDirect(block: Runnable): Disposable =
+        scheduleDirect(block, 0, TimeUnit.MILLISECONDS)
+
+    override fun scheduleDirect(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        if (!scope.isActive) return Disposables.disposed()
+        val newBlock = RxJavaPlugins.onSchedule(block)
+        return scope.launch {
+            if (delay > 0) {
+                delay(unit.toMillis(delay))
+            }
+            newBlock.run()
+        }.asDisposable()
+    }
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, schedulerJob)
+
+    override fun shutdown() {
+        scope.cancel()
+    }
+
+    private class DispatcherWorker(private val dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerJob = SupervisorJob(parentJob)
+        private val workerScope = CoroutineScope(workerJob + dispatcher)
+        private val blockChannel = Channel<SchedulerChannelTask>(Channel.UNLIMITED)
+
+        init {
+            workerJob.invokeOnCompletion {
+                blockChannel.close()
+            }
+
+            workerScope.launch {
+                while (isActive) {
+                    val task = blockChannel.receive()
+                    task.execute()
+                }
+            }
+        }
+
+        override fun isDisposed(): Boolean = !workerScope.isActive
+
+        override fun schedule(block: Runnable): Disposable =
+            schedule(block, 0, TimeUnit.MILLISECONDS)
+
+        override fun schedule(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+            if (!workerScope.isActive) return Disposables.disposed()
+
+            val newBlock = RxJavaPlugins.onSchedule(block)
+            val task = SchedulerChannelTask(delay, newBlock, dispatcher, workerJob)
+            blockChannel.offer(task)
+            return task
+        }
+
+        override fun dispose() {
+            workerScope.cancel()
+            blockChannel.close()
+        }
+    }
+}
+
+/**
+ * Represents a task to be queued sequentially on a [Channel] for a [Scheduler.Worker].
+ *
+ * Delayed tasks do not block [Channel] from processing other tasks
+ */
+private class SchedulerChannelTask(
+    private val delayMillis: Long,
+    private val block: Runnable,
+    dispatcher: CoroutineDispatcher,
+    parentJob: Job
+) : Disposable {
+    private val taskScope = CoroutineScope(dispatcher + Job(parentJob))
+    private val delayResult: Deferred<Unit>
+
+    init {
+        delayResult = taskScope.async {
+            delay(delayMillis)
+        }
+    }
+
+    fun execute() {
+        if (taskScope.isActive) {
+            if (delayResult.isCompleted) {
+                block.run()
+            } else {
+                taskScope.launch {
+                    delayResult.await()
```

Comment:
This delay logic is still wrong I think. It will not start the delay until the task is processed, and then when the delay is over it the task will not necessarily be ran exclusively with other tasks from the same worker. The delay needs to happen _before_ submitting the task to the worker queue, and then the task needs to be submitted to the queue after the delay.

The distinction is that the _delay_ can (and should) happen concurrently with other delays, and other tasks, but the actual execution of the task itself must happen in a way that doesn't "overlap" with other tasks submitted to the worker (scheduled or not).

Here's a rough diagram of the sequence that needs to happen, as I understand it:

| t (ms) | submitter1    | submitter2           | worker                        |                                                                                                                                        |
|--------|---------------|----------------------|-------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|
| 0      | schedule(foo) | schedule(bar, 1 sec) | offer(foo), delay(bar, 1 sec) | foo and bar are submitted concurrently. Worker enqueues foo immediately, but defers enqueuing bar.                                     |
| 1      |               |                      | run foo                       | Worker reads from queue, runs bar                                                                                                      |
| …      |               |                      |                               |                                                                                                                                        |
| 1000   | schedule(baz) |                      | offer(baz), offer(bar)        | After 1 sec delay, the delayed bar task is enqueued concurrently with a new task baz, but bar and baz must still be executed in order. |
| 1001   |               |                      | run baz                       |                                                                                                                                        |
| 1002   |               |                      | run bar                       |                                                                                                                                        |

Reply:
@zach-klippenstein thanks again for the explanation. sorry the expected execution order isn't clear to me from reading the docs but what you said I think makes sense.

Reply:
Done